src/dhtproto/client/legacy/internal/helper/Handshake.d

```diff
@@ -0,0 +1,211 @@
+/*******************************************************************************
+
+    Class to perform a handshake with all DHT nodes (retrying when it fails),
+    providing task helpers that block until one or all nodes are connected.
+
+    Copyright: Copyright (c) 2018 sociomantic labs GmbH.  All rights reserved
+
+    License: Boost Software License Version 1.0.  See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module dhtproto.client.legacy.internal.helper.Handshake;
+
+/// ditto
+public class DhtHandshake
+{
+    import dhtproto.client.DhtClient;
+    import dhtproto.client.legacy.internal.helper.RetryHandshake;
+
+    import ocean.io.select.EpollSelectDispatcher;
+    import ocean.task.Task;
+    import ocean.util.log.Logger;
+
+
+    /***************************************************************************
+
+        Task that suspends until all nodes are connected
+
+    ***************************************************************************/
+
+    private class AllDhtNodesConnected : Task
+    {
+        /***********************************************************************
+
+            Boolean flag that will be set to `true` once all DHT nodes have
+            been reached by the handshake
+
+        ***********************************************************************/
+
+        private bool all_dht_nodes_connected = false;
+
+        /***********************************************************************
+
+            Delegate to be called by the handshake once all DHT nodes have
+            been reached
+
+        ***********************************************************************/
+
+        private void dg ()
+        {
+            this.all_dht_nodes_connected = true;
+            this.outer.log.info("All DHT nodes connected!");
+            if (this.suspended()) this.resume();
+        }
+
+        /***********************************************************************
+
+            Runs the task
+
+        ***********************************************************************/
+
+        protected override void run ()
+        {
+            if (!this.all_dht_nodes_connected) this.suspend();
```

Comment:
Can also just use  `TaskEvent` here

Reply:
So basically `run` just contains `TaskEvent.wait();`, and the soon-to-be-renamed `dg` just contains any wanted logging plus `TaskEvent.trigger();` ... ?

Reply:
Yup

Reply:
In that sense it just occured  to me that whole handshake helper could be a task event, and not actual task - but it doesn't matter much and this way probably is more generic.

Reply:
By whole handshake helper you mean the `DhtHandshake` class?  It's not a task itself, it just exposes a couple.

If I understand you right, you mean we could just use a couple of `TaskEvent` instances under the hood and expose methods that would cause the calling task context to invoke their `wait()` ... ?

FWIW I agree that it's nicer/more flexible to expose task instances.  But I don't mind drafting an alternative design (in a separate PR) for comparison.

Reply:
Probably it could be something like:

```D
class DhtHandshake
{
    private RetryHandshake retry_handshake;

    private TaskEvent one_node_connected;

    private TaskEvent all_nodes_connected;

    public bool start (DhtClient dht, uint timeout, size_t wait_time)
    {
        verify(this.retry_handshake is null);

        void oneNodeConnectedDg (NodeItem item) { this.one_node_connected.trigger(); }

        bool timed_out = true;

        void allNodesConnectedDg ()
        {
            timed_out = false;
            this.all_nodes_connected.trigger();
        }

        this.retry_handshake =
            new RetryHandshake(theScheduler.epoll, dht,
                    &allNodesConnectedDg, &this.oneNodeConnectedDg);

        this.one_node_connected.wait();

        if (timeout > 0)
        {
            // launch some timer event that will also call
            // this.all_nodes_connected.trigger()
        }

        this.all_nodes_connected.wait();

        return timed_out;
    }
}
```

... no ... ?

Reply:
Almost what I had in mind - but you don't need to call `this.one_node_connected.wait();` there, it will be called from host ask (same for `this.all_nodes_connected.wait();`). Benefit with this approach is that you don't need to use 2 worker fibers for one/all cases, it will only suspend one host task - also allocates slightly less memory. Drawback is that it seems slightly more convoluted.

Reply:
Hmm, are you expecting that the `TaskEvent` fields will be public?  I don't see any other way to viably do what you have in mind given that they are struct instances.  And my inclination was to hide the `trigger` method at least.

FWIW I do think that on balance the existing version using tasks, while a little less efficient (in LoC as well as memory) offers a slightly more intuitive UI.  But do you think it would be worth using `TaskEvent` under the hood of the `{One,All}Node{s}Connected` tasks, in any case?

Reply:
> Hmm, are you expecting that the TaskEvent fields will be public? I 

You only need to expose public methods like `wait_one_node()` which calls event internally, actual event struct instance is of no other use anyway.

Reply:
> But do you think it would be worth using TaskEvent under the hood 

Definitely, it easier than to manually look where you have to add `if (suspended)` checks

Reply:
> You only need to expose public methods like wait_one_node() which calls event internally, actual event struct instance is of no other use anyway.

That sounds like it'll make the timeout less intuitive, though.  There's something really nice about it being in the hands of the user to say: `await(handshake.oneNode())`, `awaitOrTimeout(handshake.allNodes(), ...)`.  So, I think I'll keep the existing public API, but use `TaskEvent` under the hood of the tasks exposed.  SGTY?

Reply:
Yeah, as I have mentioned before, I am simply calling out loud available options. Current API looks reasonable.